shared/pkg/analyzer/trivy/trivy.go

```diff
@@ -42,10 +43,12 @@ type Analyzer struct {
 	localImage bool
 }
 
-func New(conf *config.Config,
-	logger *log.Entry,
-	resultChan chan job_manager.Result,
-) job_manager.Job {
+func init() {
+	job_factory.RegisterCreateJobFunc(AnalyzerName, New)
```

Comment:
Right now we have a global job factory, but I think we need a per-family job factory, the trivy analyzer should only be registering as a job within the analyzer space.

The current implementation here will run into issues when there are analyzers/scanners with the same name for example analyzer/trivy and scanner/trivy. Also there is nothing to prevent me accidentally enabling a scanner as part of an analyzer run.

I think it would make much more sense if analyzer/job looked like:

```
import (
_ "github.com/openclarity/kubeclarity/shared/pkg/analyzer/cdx_gomod"
_ "github.com/openclarity/kubeclarity/shared/pkg/analyzer/syft"
_ "github.com/openclarity/kubeclarity/shared/pkg/analyzer/trivy"
)

var JobFactory = &job_factory.Factory{}
```

trivy would look like:

```
import "github.com/openclarity/kubeclarity/shared/pkg/analyzer"

func init() {
   analyzer.JobFactory.Register(ScannerName, New) 
}
```

and then job_manager looked like:

```
func New(jobNames []string, config IsConfig, logger *logrus.Entry, jobFactory *job_factory.Factory) *Manager {
}

func (m *Manager) Run(...) {
        ...
        
     	jobs := make([]Job, len(m.jobNames))
	for i, name := range m.jobNames {
		nameToResultChan[name] = make(chan Result, 10) // nolint:gomnd
		jobs[i], err = m.jobFactory.CreateJob(name, m.config, m.logger, nameToResultChan[name])
                if err != nil {
                    log.Fatalf("unable to build list of jobs: %w": err)
                }
	}
	
	...
}
```

and cli/analyze and runtime_k8s_scanner/analyze look like:

```
manager := job_manager.New(..., analyzer.JobFactory)
```

Reply:
@sambetts-cisco I think that i've understand, but the problem that I see is that we have a cycle import here.
From your example, if trivy needs to import `analyzer` who will import `analyzer/job` for the all empty imports of the different analyzers will kick in?
unless I will do an empty import in each main that will use the job manager I'm not sure how I can solve it - I will give it a try.

Reply:
Also I have a cycle import when `job_manager` --> `job_factory` -->  `job_manager`

We might want to open a task to make a better refactor once we finalize the exact structure. meanwhile we can just call the scanner trivy and the analyzer trivy with a different name. WDYT?

Reply:
So the job_manager --> job_factory --> job_manager import cycle can be fixed by moving the Job definition to job factory, which is probably were it belongs anyway, or just moving the definition of Factory into job_manager as they are tightly coupled.

The analyzer import loop is by the directionality of our job factory registration, and we can't really get around that without having to add two imports in cli one for the job factory location and one too trigger all the registrations.

I think the way we're registering the jobs isn't recommended from a golang perspective, instead Registering to the job factory in the analyzer/job package would be better like:

```
import (
 "github.com/openclarity/kubeclarity/shared/pkg/analyzer/cdx_gomod" "github.com/openclarity/kubeclarity/shared/pkg/analyzer/syft"
 "github.com/openclarity/kubeclarity/shared/pkg/analyzer/trivy"
)

var JobFactory = &job_factory.Factory{}

func init() {
    JobFactory.Register(trivy.ScannerName, trivy.New)
    JobFactory.Register(syft.ScannerName, syft.New)
    JobFactory.Register(cdx_gomod.ScannerName, cdx_gomod.New)
}
```

alternatively we can move the factory to the analyzer package:

analyzer/analyzer.go
```
package analyzer

var JobFactory = &job_factory.Factory{}
...
```

analyzer/jobs/jobs.go
```
package jobs

import (
 _ "github.com/openclarity/kubeclarity/shared/pkg/analyzer/cdx_gomod"
 _ "github.com/openclarity/kubeclarity/shared/pkg/analyzer/syft"
 _ "github.com/openclarity/kubeclarity/shared/pkg/analyzer/trivy"
)
```

analyzer/trivy/common.go
```
import "analyzer"

func init() {
   analyzer.JobFactory.Register(ScannerName, New) 
}
```

but then cli/cmd/analyze needs to do both:

```
import _ "analyzer/jobs" // triggers all the job registrations to the job factory
import "analyzer" // to get the job factory to pass to the job_manager
```

personal opinion is that the first option is better.

Reply:
thanks @sambetts-cisco I think that https://github.com/openclarity/kubeclarity/pull/263/commits/846d884b4eb7fc121e1b97afa5b8e9f768a7844e address it